src/arch/arm/object/smc.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2021, DornerWorks Ltd.
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+#include <config.h>
+
+#ifdef CONFIG_ALLOW_SMC_CALLS
+#include <arch/object/smc.h>
+
+static exception_t performSMCCall(word_t *buffer)
+{
+    word_t i;
+    seL4_Word arg[NUM_SMC_REGS];
+    word_t *ipcBuffer;
+
+    for(i = 0; i < NUM_SMC_REGS; i++){
+        arg[i] = getSyscallArg(i, buffer);
+    }
+
+    ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
+
+    register seL4_Word r0 asm("x0") = arg[0];
+    register seL4_Word r1 asm("x1") = arg[1];
+    register seL4_Word r2 asm("x2") = arg[2];
+    register seL4_Word r3 asm("x3") = arg[3];
+    register seL4_Word r4 asm("x4") = arg[4];
+    register seL4_Word r5 asm("x5") = arg[5];
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
```

Comment:
have you considered to release the lock for the SMP kernel before the call and acquire the lock again after the call. The smc call can run arbitrary long.

Reply:
Hm, that is an important observation -- I hadn't thought about WCET on these yet at all, and without knowing more about individual calls, the kernel WCET for a system with SMC enabled is now arbitrarily high. That is even true on uni-core, but it is worse on SMP, because the kernel is blocked for all cores during that time.

This is different to other user-level behaviour that can take arbitrarily long, because this is now kernel time.

I don't think that releasing the lock while inside the kernel works in the current design: it wouldn't be a big-lock any more, and when you re-acquire the lock you would have not guarantee that the current TSC, SC, or IPC buffer still exists (some other core could have retyped them, for instance). I'm sure that could be worked-around, but it would definitely break the big-lock idea. And even if that should somehow work, it would still not solve the uni-core case.

That might throw a real spanner in the works. Do we need to turn this into some kind of more asynchronous mechanism?

Reply:
if the short term goal is, that the feature is to be merged (behind a config option) , I'm fine with the comment https://sel4.atlassian.net/browse/RFC-9?focusedCommentId=19762 from @lsf37 that for now the system designer needs to be aware of the potential pitfalls. We can look into this as a follow-up then.

Reply:
The warning message has been added here: https://github.com/seL4/seL4/pull/701/files#diff-bc9402a3a9f4e827bf5919980dd881d416c5a6225a82a380af72117a605ae299R198-R203